### DISCUSSION TITLE
Improvement about JUnit output

### DISCUSSION THREAD

**Discussion by GridexX** [oxsecurity/megalinter#1835] [Q&A]
Hello @nvuillam,
I'm an apprentice DevOps working for [r2devops](https://r2devops.io/_/r2devops-bot/mega_linter) and I'm currently improving the Megalinter job.
The execution time could be decrease by render another output format than `TAP`. `SARIF` seems to be a good alternative.   

Nevertheless some linters such as `prettier` and CSS or HTML linters are not compatible with this format, as described in the [documentation](https://oxsecurity.github.io/megalinter/latest/supported-linters/).

Moreover, we must keep the `JUnit` output format for GitLab and there aren't npm package yet to convert from `SARIF` to `JUnit` as the one used [here](https://github.com/oxsecurity/megalinter/issues/538#issuecomment-885602936) for converting from `TAP` to `JUnit`.

Have you any idea on how to keep the `JUnit` output without without converting  from `Tap`

Thanks a lot :smile:

**Comment by nvuillam** [comment]
Hi @GridexX , 

I posted on Gitlab to see if they have some ideas... there is probably a more "SAST" way than JUnit to provide results to Gitlab, because JUnit seems to be for test classes

About linters that do not manage SARIF yet... we have to find converters, or just say it's ok to not have them in results (for formatters like prettier, I don't think there is a lot of value to show the errors in Gitlab UI )

https://gitlab.com/gitlab-org/gitlab/-/issues/118496#note_1088263958

**Comment by nvuillam** [comment]
@GridexX it seems gitlab can handle a json format to accept SAST results (could be code climate format)

https://docs.gitlab.com/ee/ci/testing/code_quality.html

Code quality report widget: https://docs.gitlab.com/ee/ci/yaml/artifacts_reports.html#artifactsreportscodequality

If you implement a sarif-to-gitlab script (json to json) , that will do the job :)

**Reply by GridexX** [reply]
Good catch ! 
Codequality is free, so we can implement it in our jobs for all users.

Thanks for the help, I will look around how to implement the script. I see there is also a JSON Reporter in MegaLinter, maybe it could also do the job

**Reply by nvuillam** [reply]
JSON reporter outputs all MegaLinter structure, but it will contain linter outputs as text, so I think u'd better use SARIF reporter that outputs a big SARIF file with the output of all linters :)
I think if you implement something like a npm package sarif-to-gitlab , not only it will ease the integration of MegaLinter, but it will probably be used by lots of other tools and could become a star repo :) 
cc @thomasboni

**Reply by GridexX** [reply]
Yeah SARIF was the good option. In the meanwhile, I have updated the job and instead of creating an npm packages, there is one [GitLab snippet](https://gitlab.com/r2devops/hub/-/snippets/2402498/raw/main/sarif_to_json_megalinter.js) for now.

I will see with @thomasboni if it is worth it to take some time creating a npm package for that.

The only thing tha bother me is that `Code Climate` only display difference between the main branch and a feature branch inside a GitLab MR. 

In any case thanks a lot for you help, it was very pleasant to work with Mega_linter 😀

**Reply by nvuillam** [reply]
If you don't create a npm package, I will steal your code and do it 🦖 🦖 
Don't pass a way to package a little piece of code that can be used by thousands of users and bring attention on R2DevOps :D

**Reply by GridexX** [reply]
Have accepted the challenge :
Here is my first [npm package](https://www.npmjs.com/package/sarif-codeclimate) 😀
Thanks for the advice, enjoy a lot to write it 
Now you can steel my code 🦩

**Reply by nvuillam** [reply]
@GridexX that's great, the sooner you start in open-source, the sooner your carreer progresses :)

But where the hell is MegaLinter in your repo ? 😇  https://github.com/GridexX/sarif-codeclimate

Also: would it be possible to even improve perfs to take SARIF as stdin then output CodeClimate as stdout ? :)

So we could do some call like `cat myFile.sarif | sarif-codeclimate -`

**Reply by GridexX** [reply]
Hehe, I made a good start by working at Go2Scale !

Ow my bad about MegaLinter, I will wrote a section about it and how it is used inside the GitLab jobs :+1: 

For the perfs, the output argument is optionnal, so there is already an output as stdout. I will check how to retrieve the SARIF as stdin. Moreover, for changes, I need to update the version of the package and want to have a release tab just like this one : 
![image](https://user-images.githubusercontent.com/50552672/189062149-be4fc9c1-9816-4b56-bebc-182f354b493f.png)
How can I achieve that ? Is it related to a CHANGELOG file ?

**Reply by Kurt-von-Laven** [reply]
No, you just click on "Releases" and then "Draft a new release."

**Reply by nvuillam** [reply]
@GridexX  I was talking about using MegaLinter in your repo, so you insure the code is clean ;)

**Reply by nvuillam** [reply]
If you want you can automate the publishing of :
- beta version when you merge in main branch
- release when you create a new tag and create a github release

You can have a look at the workflows in this repo, you can probably be almost a copy-paste ;) 

https://github.com/nvuillam/node-sarif-builder/tree/main/.github/workflows

**Comment by GridexX** [comment]
By the way I also wrote the sarif-junit converter, even if it's the less SAST and a bit useless but is always display. As described [here](https://gitlab.com/gitlab-org/gitlab/-/issues/215279#note_328568219), the CodeClimate report only show in case of a comparison between the main and a feature branch.  
With this new converter, the user will see everytime a report on the MR. 

If you could review it, it would be awesome, you are giving good advices and have more experience than me :smile:

**Reply by nvuillam** [reply]
If you want you can [ask MegaLinter to have a look](https://oxsecurity.github.io/megalinter/latest/quick-start/) by installing the job, then ask us to check what MegaLinter will have not found :)
